File: test/formats/gfx/cbg_converter_test.rb
require 'lib/formats/gfx/cbg_converter'
require 'lib/virtual_file'
require 'test/test_helper'

# Unit tests for CbgConverter
class CbgConverterTest < Test::Unit::TestCase
  def test_decoding_bgr
    input_data = TestHelper.get_test_file('cbg/3')
    expected_data = TestHelper.get_test_file('cbg/3-out.png')

    file = VirtualFile.new(nil, input_data)
    CbgConverter.decode!(file, {})
    actual_data = file.data

    TestHelper.compare_image(expected_data, actual_data)
  end

  def test_decoding_grayscale
    input_data = TestHelper.get_test_file('cbg/4')
    expected_data = TestHelper.get_test_file('cbg/4-out.png')

    file = VirtualFile.new(nil, input_data)
    CbgConverter.decode!(file, {})
    actual_data = file.data

    TestHelper.compare_image(expected_data, actual_data)
  end

  def test_decoding_bgr
    input_data = TestHelper.get_test_file('cbg/ti_si_de_a1')
    expected_data = TestHelper.get_test_file('cbg/ti_si_de_a1-out.png')

    file = VirtualFile.new(nil, input_data)
    CbgConverter.decode!(file, {})
    actual_data = file.data

    TestHelper.compare_image(expected_data, actual_data)
  end
end

File: tests/formats/gfx/cbg_converter_test.cc
#include "formats/gfx/cbg_converter.h"
#include "test_support/converter_support.h"

void test_cbg_decoding(const char *input_path, const char *expected_path)
{
    Converter *converter = new CbgConverter();
    assert_decoded_image(converter, input_path, expected_path);
    delete converter;
}

int main(void)
{
    test_cbg_decoding(
        "tests/test_files/gfx/cbg/3",
        "tests/test_files/gfx/cbg/3-out.png");
    test_cbg_decoding(
        "tests/test_files/gfx/cbg/4",
        "tests/test_files/gfx/cbg/4-out.png");
    test_cbg_decoding(
        "tests/test_files/gfx/cbg/ti_si_de_a1",
        "tests/test_files/gfx/cbg/ti_si_de_a1-out.png");
    return 0;
}
